solution.cpp

```diff
@@ -8,200 +9,186 @@
 
-    // - Runtime sized arrays are a C99 feature, and a GCC and Clang extension.
-    //   Thus `bool primes[max+1];` is not standards compliant.  It should work
-    //   on Xcode (which uses Clang) and in Linux (where most people use either
-    //   GCC or Clang), but Visual Studio appears not to allow it.
-    bool * primes = (bool *) malloc( sizeof(bool) * (max+1) );
-    sieve(max+1, primes);
-
-    int ret;  // for storing return values, to check for error codes
-    //
-    ret = write_primes(max+1, primes, "input.txt", "primes.txt");
-    if (ret == -1) {
-        cout << "ERROR in `write_primes()`" << endl;
-        return 1;  // error
+int write_c (const int,
+             bool [],
+             const string & = input_file,
+             const string & = output_c);
+
+int main()
+{
+    int max = find_max();
+    
+    if (max == -1)
+    {
+        cout << "Error finding max! \n";
+        return 1; // error msg
     }
-    //
-    ret = write_composites(max+1, primes, "input.txt", "composites.txt");
-    if (ret == -1) {
-        cout << "ERROR in `write_composites()`" << endl;
-        return 1;  // error
+    else
+    {
+        cout << "Max is > " << max << endl;
     }
-
-    return 0;  // success
+    
+    bool * test_data = new bool[max];
+    
+    sieve(max, test_data);
+    
+    write_p(max, test_data);
+    
+    write_c(max, test_data);
+    
+    return 0;
 }
 
-// ----------------------------------------------------------------------------
-
-int find_max(const string& infilename) {
-    ifstream infile(infilename);
-    if (!infile.is_open())
-        return -1;  // error
-
-    int max = -1;
-
-    int n;
-    for (infile >> n; infile.good(); infile >> n) {
-        if (n > max)
-            max = n;
+int find_max (const string &input)
+{
+    int current_num, max = 0;
+    
+    if (readFile.fail())
+    {
+        return -1; // error reading file
     }
-
+    else
+    {
+        while (readFile >> current_num)
+        {
+            if (current_num > max)
+            {
+                max = current_num;
+            }
+        }
+    }
+    
+    //readFile.close();
+    
     return max;
 }
 
-void sieve(const int size, bool primes[]) {
-    // set all entries to `true`
-    for (int a = 0; a < size; a++) {
-        primes[a] = true;
+void sieve (const int test_size, bool test_data[])
+{
+    fill_n(test_data, test_size, true);     //  fill entire array to `true`
+    
+    test_data[0] = false;   //  set indices [0], [1] as `false`
+    test_data[1] = false;
+    
+    //  iterate through the indeces in the array test_data,
+    //  if an index is marked as `true`, mark all multiples of the index as `false`
+    
+    for (int p = 2; p < test_size; p++)
+    {
+        if (test_data[p])   //  if the index is true, execute. . . for loop
+        {
+            for (int m_of_p = 2; m_of_p < test_size; m_of_p++)
+            {
+                int multiple = p * m_of_p;
```

Comment:
you could avoid this line, and the `if` below it, if you set up the for loop condition and iteration a little differently -- but your code is still well written :-)